web/src/search/input/QueryInput.tsx

```diff
@@ -63,104 +52,81 @@ interface Props extends PatternTypeProps {
     hasGlobalQueryBehavior?: boolean
 }
 
-interface State {
-    /** Whether the query input is focused */
-    inputFocused: boolean
-
-    /** Whether suggestions are shown or not */
-    hideSuggestions: boolean
+interface ComponentSuggestions {
+    values: Suggestion[]
+    cursorPosition: number
+}
 
+interface State {
     /** The suggestions shown to the user */
-    suggestions: Suggestion[]
+    suggestions: ComponentSuggestions
+
+    /** All suggestions (some static and some fetched on page load) */
+    searchFilterSuggestions: SearchFilterSuggestions | null
 
-    /** Index of the currently selected suggestion (-1 if none selected) */
-    selectedSuggestion: number
+    cursorPosition: number
 }
 
-export class QueryInput extends React.Component<Props, State> {
-    private static SUGGESTIONS_QUERY_MIN_LENGTH = 2
+/**
+ * The search query and cursor position of when the last character was inserted.
+ * Cursor position is used to correctly insert the suggestion when it's selected.
+ */
+export type QueryCursorPair = [string, number]
```

Comment:
Prefer using an interface over a tuple - they are clearer and are properly inferred (tuples are not without casting)

```suggestion
export interface QueryCursorPair {
    query: string
    cursorPosition: number
}
```